### ISSUE TITLE
Whitespacing rules inside of functions

### ISSUE THREAD

**Issue by mattBrzezinski** [invenia/BlueStyle#53]
Stemming from [this](https://github.com/JuliaDiff/ChainRules.jl/pull/225#pullrequestreview-442601728) merge request I think having a discussion about whitespacing for this style guide could be beneficial.

Currently we do not have any guidelines around whitespacing inside of functions. There have been a few occasions where I've seen code formatted similarly to this:

```julia
function do_request(aws::AWSConfig, request::Request; return_headers::Bool=false)
    response = nothing
    TOO_MANY_REQUESTS = 429
    EXPIRED_ERROR_CODES = ["ExpiredToken", "ExpiredTokenException", "RequestExpired"]
    REDIRECT_ERROR_CODES = [301, 302, 303, 304, 305, 307, 308]
    THROTTLING_ERROR_CODES = [
        "Throttling",
        "ThrottlingException",
        "ThrottledException",
        "RequestThrottledException",
        "TooManyRequestsException",
        "ProvisionedThroughputExceededException",
        "LimitExceededException",
        "RequestThrottled",
        "PriorRequestNotComplete"
    ]
    if isempty(request.headers)
        request.headers = LittleDict{String, String}()
    end
    request.headers["User-Agent"] = user_agent
    request.headers["Host"] = HTTP.URI(request.url).host
    @repeat 3 try
        aws.credentials === nothing || _sign!(aws, request)
        response = @mock _http_request(aws, request)
        if response.status in REDIRECT_ERROR_CODES && HTTP.header(response, "Location") != ""
            request.url = HTTP.header(response, "Location")
            continue
        end
    catch e
        if e isa HTTP.StatusError
            e = AWSException(e)
        end
        @retry if "message" in fieldnames(typeof(e)) && occursin("Signature expired", e.message) end
        @retry if ecode(e) in EXPIRED_ERROR_CODES
            check_credentials(aws.credentials, force_refresh=true)
        end
        @delay_retry if e isa AWSException &&
            (http_status(e.cause) == TOO_MANY_REQUESTS || ecode(e) in THROTTLING_ERROR_CODES)
        end
        @retry if e isa AWSException && (
            header(e.cause, "crc32body") == "x-amz-crc32" ||
            ecode(e) in ("BadDigest", "RequestTimeout", "RequestTimeoutException")
        )
        end
    end
    if request.request_method == "HEAD"
        return Dict(response.headers)
    end
    if request.return_stream
        return request.response_stream
    end
    if request.return_raw
        return (return_headers ? (response.body, response.headers) : response.body)
    end
    mime = HTTP.header(response, "Content-Type", "")
    if isempty(mime)
        if length(response.body) > 5 && String(response.body[1:5]) == "<?xml"
            mime = "text/xml"
        end
    end
    body = String(copy(response.body))
    if occursin(r"/xml", mime)
        xml_dict_type = OrderedDict{Union{Symbol, String}, Any}
        return (return_headers ? (xml_dict(body, xml_dict_type), Dict(response.headers)) : xml_dict(body, xml_dict_type))
    end
    if occursin(r"json$", mime)
        if isempty(response.body)
            return (return_headers ? (nothing, response.headers) : nothing)
        end
        info = request.ordered_json_dict ? JSON.parse(body, dicttype=OrderedDict) : JSON.parse(body)
        return (return_headers ? (info, Dict(response.headers)) : info)
    end
    return (return_headers ? (response.body, response.headers) : response.body)
end
``` 

Personally I find reading something like this to be very difficult and at a quick glance it's hard to tell everything that is happening in it. I'd like to propose a few additions to the styling guide

1) Grouping similar one-line statements together
```julia
# Yes
response = nothing
TOO_MANY_REQUESTS = 429
EXPIRED_ERROR_CODES = ["ExpiredToken", "ExpiredTokenException", "RequestExpired"]
REDIRECT_ERROR_CODES = [301, 302, 303, 304, 305, 307, 308]

# No
response = nothing

TOO_MANY_REQUESTS = 429

EXPIRED_ERROR_CODES = ["ExpiredToken", "ExpiredTokenException", "RequestExpired"]

REDIRECT_ERROR_CODES = [301, 302, 303, 304, 305, 307, 308]
```
2) Do NOT group multi-line statements together
```julia
# Yes
if request.request_method == "HEAD"
    return Dict(response.headers)
end

if request.return_stream
    return request.response_stream
end

# No
if request.request_method == "HEAD"
    return Dict(response.headers)
end
if request.return_stream
    return request.response_stream
end
```

3) Do NOT whitespace nested control flow statements
```julia
# Yes
if isempty(mime)
    if length(response.body) > 5 && String(response.body[1:5]) == "<?xml"
        mime = "text/xml"
    end
end

# No
if isempty(mime)
    
    if length(response.body) > 5 && String(response.body[1:5]) == "<?xml"
        mime = "text/xml"
    end

end
```

4) Optional line break between one-line statements for `return`
```julia
# Yes, preferable
if occursin(r"json$", mime)
    if isempty(response.body)
        return (return_headers ? (nothing, response.headers) : nothing)
    end

    info = request.ordered_json_dict ? JSON.parse(body, dicttype=OrderedDict) : JSON.parse(body)
    return (return_headers ? (info, Dict(response.headers)) : info)
end

# Yes, acceptable
if occursin(r"json$", mime)
    if isempty(response.body)
        return (return_headers ? (nothing, response.headers) : nothing)
    end

    info = request.ordered_json_dict ? JSON.parse(body, dicttype=OrderedDict) : JSON.parse(body)

    return (return_headers ? (info, Dict(response.headers)) : info)
end

# Yes, preferable
if occursin(r"json$", mime)
    if isempty(response.body)
        return (return_headers ? (nothing, response.headers) : nothing)
    end
    
    info = Dict()
    info["foo"] = "bar"
    delete!(info, "foo")
    info = request.ordered_json_dict ? JSON.parse(body, dicttype=OrderedDict) : JSON.parse(body)

    return (return_headers ? (info, Dict(response.headers)) : info)
end

# Yes, acceptable
if occursin(r"json$", mime)
    if isempty(response.body)
        return (return_headers ? (nothing, response.headers) : nothing)
    end
    
    info = Dict()
    info["foo"] = "bar"
    delete!(info, "foo")
    info = request.ordered_json_dict ? JSON.parse(body, dicttype=OrderedDict) : JSON.parse(body)
    return (return_headers ? (info, Dict(response.headers)) : info)
end
```

5) Line break between control flow statement and `return`
```julia
# Yes
if isempty(mime)
    if length(response.body) > 5 && String(response.body[1:5]) == "<?xml"
        mime = "text/xml"
    end
end

return mime

# No
if isempty(mime)
    if length(response.body) > 5 && String(response.body[1:5]) == "<?xml"
        mime = "text/xml"
    end
end
return mime
```

**Comment by mattBrzezinski** [comment]
cc: @nickrobinson251 @oxinabox @fchorney @nicoleepp

**Comment by fchorney** [comment]
I agree with all these points. I'm all for saving some vertical space when we can, but I definitely prefer when logical sections are split up.

**Comment by nickrobinson251** [comment]
Mostly seems fine to me. 

Can we say "blank line" rather than "whitespace"? it seems clearer to me. 

Can we simplify 4 to "a blank line before `return` is optional"?

I'm not sure I see any benefit to 5, given we have 4. 

And i'm not sure we need such a big example to show 4 (we should try to keep examples minimal and only where they clarify the point, so the guide doesn't become verbose).

**Comment by mattBrzezinski** [comment]
> Can we say "blank line" rather than "whitespace"? it seems clearer to me.

Yeah, when I make the PR I can change the language to be more clear.
 
> Can we state things in the positive when possible? e.g. 2 can be written as something like "Use blank-lines to separate different multi-line blocks."

:+1: 

> Can we simplify 4 to "a blank line before `return` is optional"?

Optional but preferred?

> I'm not sure I see any benefit to 5, given we have 4.

I just wanted to handle this case as well, but it can be unified together.

> And i'm not sure we need such a big example to show 4 (we should try to keep examples minimal and only where they clarify the point, so the guide doesn't become verbose).

Agreed, I just wanted to be verbose in these examples for clarity. I'll simplify it when making the PR.

**Comment by nicoleepp** [comment]
For 3, there are cases where the if is long and hard to read, and adding the blank line immediately after improves readability.

I don't really see the benefit of 4 unless we phrase it as "A blank line before a return statement is optional but preferred". Also vote to get rid of 5

For 2, if the `if` statements are logically grouped together I don't see the point of the blank line. If not logically of the same grouping then yes, blank line.

I'm 100% behind 1

**Comment by omus** [comment]
I can get behind this. My only real gripe is that for 2 the multiple return statements could easily be turned into an `elseif` which limits the usefulness of those examples

**Comment by TylerLoewen** [comment]
I think it is also worth bringing up blank lines directly following function definitions. For example I have seen both of the following cases used.

```Julia
# No blank line
function print(data::string)
    println(data)
end
```

```Julia
# Blank line
function print(data::string)

    println(data)
end
```

This could similarly include blank lines after `return` statements before the function's closing `end` statement. In my opinion having no blank line is more readable for functions with only a few lines.

**Comment by mattBrzezinski** [comment]
Closed in #54
